### PR TITLE
Set hb_buffer_t to use array_t.reverse().

### PR DESCRIPTION
**Pull request by garretrieger**

### CODE REVIEW

File: src/hb-array.hh [hunk 1]
```diff
@@ -173,16 +173,25 @@ struct hb_array_t : hb_iter_with_fallback_t<hb_array_t<Type>, Type&>
...
-    int lhs = 0;
-    while (rhs > lhs)
-    {
-      Type value_rhs = arrayZ[rhs];
-      Type value_lhs = arrayZ[lhs];
-      arrayZ[rhs] = value_lhs;
-      arrayZ[lhs] = value_rhs;
-      rhs--;
-      lhs++;
+    return reverse (0, length);
```
Comment (behdad) [inline src/hb-array.hh]: Make start/end have defaults so you don't need separate override.

File: src/hb-array.hh [hunk 2]
```diff
@@ -173,16 +173,25 @@ struct hb_array_t : hb_iter_with_fallback_t<hb_array_t<Type>, Type&>
...
+
+  /*
+   * Reverse the order of items in this array in the range [start, end).
+   */
+  void reverse (unsigned start, unsigned end)
+  {
+    start = hb_min (start, length);
+    end = hb_min (end, length);
+
+    if (end - start < 2)
```
Comment (behdad) [inline src/hb-array.hh]: Problematic if end < start.

File: src/hb-buffer.cc [hunk 3]
```diff
@@ -448,27 +448,15 @@ void
...
 
-  for (i = start, j = end - 1; i < j; i++, j--) {
-    hb_glyph_info_t t;
-
-    t = info[i];
-    info[i] = info[j];
-    info[j] = t;
-  }
+  hb_array_t<hb_glyph_info_t> info_array(info, len);
+  info_array.reverse (start, end);
```
Comment (behdad) [inline src/hb-buffer.cc]: Just `hb_array (info, len).reverse (start, end);`. Same below.

File: src/hb-array.hh (hunk 1 above)
Comment (garretrieger) [reply src/hb-array.hh]: Done.

File: src/hb-buffer.cc (hunk 3 above)
Comment (garretrieger) [reply src/hb-buffer.cc]: Done.

File: src/hb-array.hh (hunk 2 above)
Comment (garretrieger) [reply src/hb-array.hh]: Done.

File: src/hb-array.hh [hunk 4]
```diff
@@ -171,18 +171,21 @@ struct hb_array_t : hb_iter_with_fallback_t<hb_array_t<Type>, Type&>
 
   unsigned int get_size () const { return length * this->get_item_size (); }
 
-  void reverse ()
+  /*
+   * Reverse the order of items in this array in the range [start, end).
+   */
+  void reverse (unsigned start = 0, unsigned end = -1)
```
Comment (behdad) [inline src/hb-array.hh]: Maybe use `UINT_MAX` or `(unsigned) -1`.  I don't know which one we're using these days.

File: src/hb-buffer.cc [hunk 5]
```diff
@@ -448,27 +448,13 @@ void
...
     return;
 
-  for (i = start, j = end - 1; i < j; i++, j--) {
-    hb_glyph_info_t t;
-
-    t = info[i];
-    info[i] = info[j];
-    info[j] = t;
-  }
+  hb_array_t<hb_glyph_info_t> (info, len).reverse (start, end);
```
Comment (behdad) [inline src/hb-buffer.cc]: Again.  I meant `hb_array (info, len).reverse (start, end);`.  `hb_array` is a function that deduces the template type for you so you don't have to type it.

**Pull request merged**